### DISCUSSION TITLE
Dialog does not return returnValue correctly on Vue

### DISCUSSION THREAD

**Discussion by mai1202** [material-components/material-web#5369] [Q&A] [answered]
main.js
```
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'
import '@material/web/button/text-button.js';
import "@material/web/dialog/dialog.js";

createApp(App).use(router).mount('#app')
```
test.vue
```<template>
    <md-dialog open>
        <form slot="content" id="form-id" method="dialog">...</form>
        <div slot="actions">
            <md-text-button form="form-id" value="cancel">Cancel</md-text-button>
            <md-text-button form="form-id" value="ok">Ok</md-text-button>
        </div>
    </md-dialog>
</template>

<script setup>
import { onMounted } from "vue";
let d;
onMounted(()=> {
    d = document.querySelector("md-dialog");
    d.returnValue = "default";
    d.addEventListener('close', () => {
        console.log(d.returnValue);
    });
})
</script>
```
When I close the dialog, the console will output “default”. If remove `d.returnValue = "default";`, it will output a empty string.

Is it a bug or my wrong?

**Comment by asyncliz** [comment]
How are you closing the dialog, and what do you expect the return value to be?

If you cancel the dialog by pressing the escape key or clicking the background, the `returnValue` property is not changed, so it would be "default" since you set it to that, or its default empty string when unset.

[When I try using the dialog](https://lit.dev/playground/#gist=83dbd83e2b46d988e080ca847543d4bc) it correctly reports "default" when canceled and either "cancel" or "ok" when the buttons are pressed.

**Comment by mai1202** [comment]
> How are you closing the dialog, and what do you expect the return value to be?
> 
> If you cancel the dialog by pressing the escape key or clicking the background, the `returnValue` property is not changed, so it would be "default" since you set it to that, or its default empty string when unset.
> 
> [When I try using the dialog](https://lit.dev/playground/#gist=83dbd83e2b46d988e080ca847543d4bc) it correctly reports "default" when canceled and either "cancel" or "ok" when the buttons are pressed.

@asyncLiz This situation seems to only occur on Vue. Example: https://github.com/mai1202/mw-test

**Reply by asyncliz** [reply]
Can you provide an online reproduction with something like [Stackblitz](https://stackblitz.com/fork/vue)?

**Reply by mai1202** [reply]
@asyncLiz When I ran it on StackBlitz, it returned 
```
Error in /turbo_modules/@material/web@1.1.1/button/text-button.js (10:25)
Unexpected token 'export'
```
Could you run it locally?

**Reply by asyncliz** [reply]
Unfortunately it's a bit hard to download and run projects on my work laptop.

Is it possible you need to wait for the custom element to register before setting the property? Vue's `onMounted()` callback may be too early.

```html
<script setup>
import { onMounted } from "vue";
let d;
onMounted(()=> {
    customElements.whenDefined('md-dialog').then(() => {
        d = document.querySelector("md-dialog");
        d.returnValue = "default";
        d.addEventListener('close', () => {
            console.log(d.returnValue);
        });
    });
})
</script>
```

Can you also let us know what you expect the return value to be, and how you are closing the dialog?

**Reply by mai1202** [reply]
@asyncLiz I tried changing to the code you gave me but it doesn't work.

I think maybe it's not a running time problem. I've tried running it with setTimeout() before, but it still didn't work.

Regarding the return value, I expect it to return “cancel” when I click Cancel and “ok” when I click Ok. And this is a video of it operating on Vue: https://github.com/material-components/material-web/assets/63698132/9ddb835e-c6ae-4c32-9dc5-c1c1a7d728a6

**Reply by asyncliz** [reply]
The best way we could help figure this out is if we can isolate the problem.

First, can you try installing `@material/web@nightly` and see if the issue is still present? Let's make sure it didn't get fixed already.

Next, let's try to create a reproduction without Vue. If we can or cannot reproduce it outside of Vue, that will help us narrow it down to something specific with Vue or the dialog element.

**Comment by mai1202** [comment]
Any progress here?

**Reply by asyncliz** [reply]
No progress updates to report. Thanks for your patience! We're stretched pretty thin right now

**Comment by mai1202** [comment]
I've upgraded @material/web to the latest version.

Then I tried using without Vue, as this:https://github.com/mai1202/mw-test/blob/main/without-vue/index.html . It works.
![image](https://github.com/material-components/material-web/assets/63698132/4859175a-e198-41f6-b2e9-0d05ab355089)

But on Vue, it still did not work.
![image](https://github.com/material-components/material-web/assets/63698132/b65c0bf3-4dd7-4aa3-a902-2c99d50c447b)

**Reply by asyncliz** [reply]
I was able to run your app, and it looks like the `value` attributes are removed from the text buttons.

<img width="445" alt="Screenshot 2024-01-22 at 10 53 40 AM" src="https://github.com/material-components/material-web/assets/34382620/4b7576e1-ccdf-4ef1-9d56-c1cb454b092a">

This means the dialog doesn't update its `returnValue` since it can't find a `value` attribute on the button.

Digging in deeper, it looks like Vue will decide to try and set the property `button.value = 'cancel'` instead of `button.setAttribute('value', 'cancel')`. Our property doesn't reflect to the attribute, and then things break.

We can fix this on our end by making the button's `value` attribute reflects.

**Reply by asyncliz** [reply, marked as answer]
I filed https://github.com/material-components/material-web/issues/5409 to fix it. Should have a fix out shortly, it's pretty straightforward. You can install with `@material/web@nightly` tomorrow.

**Reply by mai1202** [reply]
Thank you so much!
